### DISCUSSION TITLE
Issue getting NMEA data in from serial device

### DISCUSSION THREAD

**Discussion by PrincessTwitchy** [jvde-github/AIS-catcher#143] [General]
Hello, I have been a AIS-catcher user for a bit now with my SDR dongles, and I recently bought a NMEA capable transceiver for the boat.
I figured I try AIS-catcher with serial, but have been having trouble getting the serial mode to decode any NMEA messages. I am using todays v .50 x64 windows binaries. I am using "-e 38400 COM3" It opens the port, and I confirmed that it is using the port, as Putty can no longer use it while AIS-catcher is running, but I am not decoding any NMEA messages in AIS-catcher. I am pretty sure my messages are ok. example from my seral device !AIVDM,1,1,,A,B5`OG1h006T24=3u0CSQ3wr5kP06,0*30 Is there a way to see the raw input that AIS-catcher is seeing? Not sure if the issue is happening in the serial side, or in the decode side of things, or if there is something simple I missed.  
  
[output.txt](https://github.com/jvde-github/AIS-catcher/files/12255268/output.txt)

**Comment by jvde-github** [comment]
Hello, that looks ok to me: the command line, NMEA and output. Another way to check whether you are reading from the right device is to run AIS-catcher and to disconnect the USB cable. The program should then stop.

I have added an option so that you indeed can print the input from the serial device on the command line (`-ge print on`). As you are using Windows you can download the latest edge version [here](https://github.com/jvde-github/AIS-catcher/releases/tag/Edge).
Below the output from my run with a Comar connected to my Windows PC. The lines with `>>` indicate the raw output received from the device.

As a final question, are you using particular special settings for the serial connection in Putty? I have not implemented all options here (only one-stop-bit and no-parity). 
You can find it here in Putty:

![image](https://github.com/jvde-github/AIS-catcher/assets/52420030/60aa550e-5c02-4163-94a6-22e5abad00d7)



```
AIS-catcher (build Aug  4 2023) Edge-2-gf6d7c30
(C) Copyright 2021-2023 jvde-github and other contributors
This is free software; see the source for copying conditions.There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Server: no plugin files found in directory.
UDP: open socket for host: 127.0.0.1, port: 10110, filter: OFF, JSON: OFF
HTML Server running at port 8100
Device    : SERIAL
Settings  : rate 288K format TXT baudrate 38400 port COM3
Model #0 -> 1: [NMEA input] nmea_refresh OFF crc_check OFF
[NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
[NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
[NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
>> !AIVDM,1,1,,B,13bVm>0P00PCQ=8Mksnf4?v62D6O,0*7E
!AIVDM,1,1,,B,13bVm>0P00PCQ=8Mksnf4?v62D6O,0*7E
[NMEA input #0-0]                       received: 1 msgs, total: 1 msgs, rate: 0.0995294 msg/s
[NMEA input #0-0]                       received: 0 msgs, total: 1 msgs, rate: 0 msg/s
>> !AIVDM,1,1,,B,13bVm>0P00PCQ<lMksp>4?vd28=V,0*51
!AIVDM,1,1,,B,13bVm>0P00PCQ<lMksp>4?vd28=V,0*51
[NMEA input #0-0]                       received: 1 msgs, total: 2 msgs, rate: 0.0996002 msg/s
```

**Reply by PrincessTwitchy** [reply]
Thanks Jasper,  I went ahead and tried your latest edge version, with the -ge argument, and it looks like I am indeed not getting anything in on COM3 with AIS catcher. It opens the port successfully, but just doesn't receive any data. I have confirmed that the issue is in the serial portion of AIS catcher, as I was successful in using NmeaRouter to open the serial port, and route the data to a UDP port in AIS Catcher. I attached my working Nmea router setup, (also attached my putty settings just to be through) 

(Edit: I should add that It is a USB to serial device in windows, not sure if that has any affect on the landscape)

![Capture](https://github.com/jvde-github/AIS-catcher/assets/29732767/e76c8aab-e42f-4f0d-b2f0-f6a8ae45e33a)


![putty](https://github.com/jvde-github/AIS-catcher/assets/29732767/7620f4ee-96a1-47c9-b108-41c1f38c4583)

> C:\Users\AISPC\Desktop\ais catcher edge>AIS-catcher  -n -v 10 -e 38400 COM3 -ge print on -u 127.0.0.1 10110 -N 8100 PLUGIN_DIR plugins
> AIS-catcher (build Aug  4 2023) Edge-42-g9331119
> (C) Copyright 2021-2023 jvde-github and other contributors
> This is free software; see the source for copying conditions.There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> Server: adding plugin (JS): plugins\dimension_in_feet.pjs
> Server: adding plugin (JS): plugins\google_option.pjs
> Server: adding plugin (JS): plugins\latlon_in_dms.pjs
> Server: adding plugin (JS): plugins\maps_with_no_labels.pjs
> Server: adding plugin (JS): plugins\marinetraffic_option.pjs
> Server: adding plugin (JS): plugins\shipxplorer_option.pjs
> Server: adding plugin (JS): plugins\vesselfinder_option.pjs
> UDP: open socket for host: 127.0.0.1, port: 10110, filter: OFF, JSON: OFF
> HTML Server running at port 8100
> Device    : SERIAL
> Settings  : rate 288K format TXT baudrate 38400 port COM3 print ON
> Model #0 -> 1: [NMEA input] nmea_refresh OFF crc_check OFF
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s
> [NMEA input #0-0]                       received: 0 msgs, total: 0 msgs, rate: 0 msg/s

**Comment by jvde-github** [comment]
Thanks for being so systematic and for flagging the issue! Your device setup looks ok (1 stop bit, no parity) and programs like Putty seem to work (you have 34800 for Putty but assume it works for both 38400 and 34800). 

I have reworked the code a bit (automatically available as [Edge binaries](https://github.com/jvde-github/AIS-catcher/releases/tag/Edge)) to initialize the port slightly differently. Perhaps this resolves it. It might be the DTR or RTS control setting for particular devices...

I have tested though only on a Comar R400N (connected via USB available on COM3) with a Windows 11 PC runing Catcher 32bit and 64bit. if this did not fix it, anything specific for your system setup that could provide insights for further testing?

Thank you again, Jasper

**Reply by PrincessTwitchy** [reply]
Thanks Jasper, that did the trick! I am now getting serial data directly into AIS catcher! Thanks for the great work on such an amazing program! I'm glad I was able to help out with troubleshooting.

Thanks,
Twitchy

**Reply by jvde-github** [reply]
Great to hear! Thank you for reporting and testing. Without that it would likely not be found and fixed. Glad that it is now working. Thanks, Jasper
